### PR TITLE
--options-file command line option: support passing additional arguments via a text file

### PR DESCRIPTION
**Pull request by mosherubin**
In my work with Frida I have generated many hundreds of -i / -a command line arguments.  Running under Windows I have hit the maximal command line size, and have even encountered the same problem with Linux.

This feature allow passing an unlimited number of arguments via text files.  These text files are tokenized correctly and accumulated, eventually being passed to `parser.parse_args` for parsing and processing.

The feature opens the door to automated generated files of hundreds or even thousands of arguments, typical of modern day usage of tools like decompilers (e.g., IDA or Ghidra) and frida-trace.

### CODE REVIEW

File: frida_tools/application.py [hunk 1]
```diff
@@ -445,6 +451,47 @@ def work():
...
+                if i < len(arglist)-1:
+                    return i
+                else:
+                    # No argument following the option, error
+                    parser.error ('No argument given for -O option')
+        return -1
+
+    def insert_new_args_in_list(self, args, offset, new_arg_text):
+        new_args = shlex.split(new_arg_text)
+        new_args_list = args[:offset] + new_args + args[offset+2:]
```
Comment (oleavr) [inline frida_tools/application.py]:
```suggestion
        new_args_list = args[:offset] + new_args + args[offset + 2:]
```

File: frida_tools/application.py [hunk 2]
```diff
@@ -445,6 +451,47 @@ def work():
...
+
+            real_args = self.insert_new_args_in_list(real_args, offset, new_arg_text)
+            files_processed.add(file_path)
+
+        return real_args
+
+    def find_arg_file_offset(self, arglist, parser):
+        for i in range(len(arglist)):
+            if (arglist[i] == '-O') or (arglist[i] == '--options-file'):
+                if i < len(arglist)-1:
```
Comment (oleavr) [inline frida_tools/application.py]:
```suggestion
                if i < len(arglist) - 1:
```

File: frida_tools/application.py [hunk 3]
```diff
@@ -445,6 +451,47 @@ def work():
...
+                    return i
+                else:
+                    # No argument following the option, error
```
Comment (oleavr) [inline frida_tools/application.py]: Redundant comment.

File: frida_tools/application.py [hunk 4]
```diff
@@ -445,6 +451,47 @@ def work():
...
+                    parser.error ('No argument given for -O option')
```
Comment (oleavr) [inline frida_tools/application.py]:
```suggestion
                    parser.error("No argument given for -O option")
```

File: frida_tools/application.py [hunk 5]
```diff
@@ -445,6 +451,47 @@ def work():
...
+            file_path = real_args[offset+1]
+            
+            if file_path in files_processed:
+                parser.error('File "' + file_path + '" given twice as -O argument')
+
+            if os.path.isfile(file_path):
+                with codecs.open(file_path, 'r', 'utf-8') as f:
+                  new_arg_text = f.read()
+            else:
+                parser.error('File "' + file_path + '" following -O option is not a valid file')
```
Comment (oleavr) [inline frida_tools/application.py]: Strings should be double-quoted. Better to use `.format()` here.

File: frida_tools/application.py [hunk 6]
```diff
@@ -445,6 +451,47 @@ def work():
 
         return result[0]
 
+    def return_real_args(self, parser):
+        real_args = sys.argv[1:]
+        performPass = True
+        files_processed = set()
+
+        while (performPass):
```
Comment (oleavr) [inline frida_tools/application.py]:
```suggestion
        while performPass:
```

File: frida_tools/application.py [hunk 7]
```diff
@@ -445,6 +451,47 @@ def work():
 
         return result[0]
 
+    def return_real_args(self, parser):
+        real_args = sys.argv[1:]
+        performPass = True
```
Comment (oleavr) [inline frida_tools/application.py]: Variable names should not be camelCase, but snake_case. (We're following PEP-8 with the addition of double- vs. single-quotes for strings vs. enums.)

File: frida_tools/application.py [hunk 8]
```diff
@@ -445,6 +451,47 @@ def work():
 
         return result[0]
 
+    def return_real_args(self, parser):
```
Comment (oleavr) [inline frida_tools/application.py]: It seems like these helper functions don't need to access `self`, so they could be moved outside the class (below it).

File: frida_tools/application.py [hunk 9]
```diff
@@ -445,6 +451,47 @@ def work():
...
+                continue
+
+            file_path = real_args[offset+1]
+            
+            if file_path in files_processed:
+                parser.error('File "' + file_path + '" given twice as -O argument')
+
+            if os.path.isfile(file_path):
+                with codecs.open(file_path, 'r', 'utf-8') as f:
+                  new_arg_text = f.read()
```
Comment (oleavr) [inline frida_tools/application.py]: Inconsistent indentation.

File: frida_tools/application.py [hunk 10]
```diff
@@ -102,9 +104,13 @@ def store_target(option, opt_str, target_value, parser, target_type, *args, **kw
             parser.add_option("--debug", help="enable the Node.js compatible script debugger",
                 action='store_true', dest="enable_debugger", default=False)
 
+        parser.add_option("-O", "--options-file", help="text file containing additional command line options",
+                metavar="FILE", type='string', action='store')
+
         self._add_options(parser)
 
-        (options, args) = parser.parse_args()
+        real_args = self.return_real_args (parser)
```
Comment (oleavr) [inline frida_tools/application.py]:
```suggestion
        real_args = self.return_real_args(parser)
```

Perhaps naming this something like `compute_real_args` would be clearer?

File: frida_tools/application.py (hunk 1 above)
Comment (mosherubin) [reply frida_tools/application.py]: Done.  What coding rule is it to pad the '+' operator with spaces?

File: frida_tools/application.py (hunk 2 above)
Comment (mosherubin) [reply frida_tools/application.py]: Done

File: frida_tools/application.py (hunk 3 above)
Comment (mosherubin) [reply frida_tools/application.py]: Done

File: frida_tools/application.py (hunk 4 above)
Comment (mosherubin) [reply frida_tools/application.py]: Done

File: frida_tools/application.py (hunk 5 above)
Comment (mosherubin) [reply frida_tools/application.py]: Done

File: frida_tools/application.py (hunk 6 above)
Comment (mosherubin) [reply frida_tools/application.py]: Done

File: frida_tools/application.py (hunk 7 above)
Comment (mosherubin) [reply frida_tools/application.py]: Done

File: frida_tools/application.py (hunk 8 above)
Comment (mosherubin) [reply frida_tools/application.py]: Good catch -- done!

File: frida_tools/application.py (hunk 9 above)
Comment (mosherubin) [reply frida_tools/application.py]: Done

File: frida_tools/application.py (hunk 10 above)
Comment (mosherubin) [reply frida_tools/application.py]: Done

File: frida_tools/application.py (hunk 1 above)
Comment (oleavr) [reply frida_tools/application.py]: It's defined in [PEP-8](https://www.python.org/dev/peps/pep-0008/), which our style is based on. (Our only addition is the double quotes vs single quotes rule based on string vs enum.)

File: frida_tools/application.py [hunk 11]
```diff
@@ -445,6 +451,46 @@ def work():
...
+        else:
+            parser.error("File '{}' following -O option is not a valid file".format(file_path))
+
+        real_args = insert_new_args_in_list(real_args, offset, new_arg_text)
+        files_processed.add(file_path)
+
+    return real_args
+
+def find_arg_file_offset(arglist, parser):
+    for i in range(len(arglist)):
```
Comment (oleavr) [inline frida_tools/application.py]:
```suggestion
    for i, arg in enumerate(arglist):
```

File: frida_tools/application.py [hunk 12]
```diff
@@ -445,6 +451,46 @@ def work():
...
+        if (arglist[i] == '-O') or (arglist[i] == '--options-file'):
```
Comment (oleavr) [inline frida_tools/application.py]:
```suggestion
        if arg in ("-O", "--options-file"):
```

File: frida_tools/tracer.py [hunk 13]
```diff
@@ -347,6 +347,22 @@ def _create_resolver_script(self):
...
+        f = res[0];
+    } else {
+        m = (res[0] === '') ? '*' : res[0];
+        f = (res[1] === '') ? '*' : res[1];
+    }
+    return {
+        module: m,
+        function: f
+    };
+}
```
Comment (oleavr) [inline frida_tools/tracer.py]: Rebasing should get rid of this unrelated change.

File: frida_tools/application.py [hunk 14]
```diff
@@ -18,6 +19,7 @@
 import colorama
 from colorama import Cursor, Fore, Style
 import frida
+import shlex
```
Comment (oleavr) [inline frida_tools/application.py]: This should be moved up to the first group of imports. The first group is for imports from the Python standard library, the second is for external imports.

File: frida_tools/application.py [hunk 15]
```diff
@@ -445,6 +451,46 @@ def work():
...
         return result[0]
 
+def compute_real_args(parser):
+    real_args = sys.argv[1:]
+    perform_pass = True
+    files_processed = set()
+
+    while perform_pass:
+        offset = find_arg_file_offset(real_args, parser)
+        if (offset < 0):
```
[the remark this answers is not in the document]
Comment (mosherubin) [reply frida_tools/application.py]: Done.

File: frida_tools/application.py [hunk 16]
```diff
@@ -445,6 +451,46 @@ def work():
...
+            perform_pass = False
+            continue
+
+        file_path = real_args[offset+1]
+
+        if file_path in files_processed:
+            parser.error("File '{}' given twice as -O argument".format (file_path))
```
[the remark this answers is not in the document]
Comment (mosherubin) [reply frida_tools/application.py]: Done.

File: frida_tools/application.py [hunk 17]
```diff
@@ -445,6 +451,46 @@ def work():
...
+    perform_pass = True
+    files_processed = set()
+
+    while perform_pass:
+        offset = find_arg_file_offset(real_args, parser)
+        if (offset < 0):
+            perform_pass = False
+            continue
+
+        file_path = real_args[offset+1]
```
[the remark this answers is not in the document]
Comment (mosherubin) [reply frida_tools/application.py]: Done.

File: frida_tools/application.py (hunk 11 above)
Comment (mosherubin) [reply frida_tools/application.py]: New Pythonic idiom for me :-).  Done.

File: frida_tools/application.py (hunk 12 above)
Comment (mosherubin) [reply frida_tools/application.py]: Elegant, done.

File: frida_tools/application.py (hunk 14 above)
Comment (mosherubin) [reply frida_tools/application.py]: Done.

File: frida_tools/tracer.py (hunk 13 above)
Comment (mosherubin) [reply frida_tools/tracer.py]: I'm not confident enough to know how to rebase these lines out, and I don't want to mess things up.  Can you rebase this for me (and explain what you did)?

Comment (mosherubin) [reply frida_tools/tracer.py]: Reading up on rebasing, I moved to the "options_file_impl" branch and ran:

```
D:\Private\Dev\Open-Source-Projects\frida\frida-tools>git rebase master
Current branch options_file_impl is up to date.
```

`git status` shows no changes to `tracer.py`.  Please walk me through what commands I need to run to get the desired results.

Comment (oleavr) [reply frida_tools/tracer.py]: Do `git checkout master` and then `git pull`, so your local `master` is up-to-date with the upstream one. Then after that `git rebase`. (But make a backup first.)

Comment (mosherubin) [reply frida_tools/tracer.py]: OK, did this. `git status` doesn't show any modifications to `tracer.py`, but I assume that the next time I commit/push the `options_file_impl` branch, it will fix itself.

Comment (oleavr) [reply frida_tools/tracer.py]: After that you can `git push -f` to overwrite the remote branch, which will update the PR.

Comment (mosherubin) [reply frida_tools/tracer.py]: That was just what I needed <g>!

```
D:\Private\Dev\Open-Source-Projects\frida\frida-tools>git branch
  init_session_impl
  master
* options_file_impl
  user_data_impl

D:\Private\Dev\Open-Source-Projects\frida\frida-tools>git push -f
fatal: AggregateException encountered.
   One or more errors occurred.
Username for 'https://github.com': xxxxxxxxxx
Password for 'https://xxxxxxxxxx@github.com':
Everything up-to-date

D:\Private\Dev\Open-Source-Projects\frida\frida-tools>
```
If the remote was indeed updated, then the problem should be solved.  Please tell me if it is.

Comment (oleavr) [reply frida_tools/tracer.py]: Great, thank you! (Sorry for the delay here, I should be back on track soon 😅)

**Comment by oleavr** [comment]
Landed with minor tweaks as 1d6c6689d2c89cc079457f1d2d7ccbe049155035. Thanks for this awesome feature! :heart:

**Comment by oleavr** [comment]
Released in 5.4.0 :tada:
